Signal/src/ViewControllers/ContactViewController.swift

```diff
@@ -0,0 +1,686 @@
+        SwiftAssertIsOnMainThread(#function)
+
+        return phoneNumbersForContact().filter({ (phoneNumber) -> Bool in
+            return contactsManager.hasSignalAccount(forRecipientId: phoneNumber)
+        })
+    }
+
+    private func systemContactsForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        return phoneNumbersForContact().filter({ (phoneNumber) -> Bool in
+            return contactsManager.allContactsMap[phoneNumber] != nil
+        })
+    }
+
+    private func phoneNumbersForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        var result = [String]()
+        if let phoneNumbers = contact.phoneNumbers {
+            for phoneNumber in phoneNumbers {
+                result.append(phoneNumber.phoneNumber)
+            }
+        }
+        return result
+    }
+
+    private func tryToDetermineModeRetry() {
+        // Try again after a minute.
+        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 60.0) { [weak self] in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        }
+    }
+
+    private func updateContent() {
+        SwiftAssertIsOnMainThread(#function)
+
+        let rootView = self.scrollView
+
+        for subview in rootView.subviews {
+            subview.removeFromSuperview()
+        }
+
+        // TODO: The design calls for no navigation bar, just a back button.
+        let topView = UIView.container()
+        topView.backgroundColor = UIColor(rgbHex: 0xefeff4)
+        topView.preservesSuperviewLayoutMargins = true
+        rootView.addSubview(topView)
+        topView.autoPinEdge(toSuperviewEdge: .top)
+        topView.autoPinEdge(.left, to: .left, of: self.view, withOffset: 0)
+        topView.autoPinEdge(.right, to: .right, of: self.view, withOffset: 0)
+
+        // TODO: Use actual avatar.
+        let avatarSize = CGFloat(100)
+        let avatarView = UIView.container()
+        avatarView.backgroundColor = UIColor.ows_materialBlue
+        avatarView.layer.cornerRadius = avatarSize * 0.5
+        topView.addSubview(avatarView)
+        avatarView.autoPin(toTopLayoutGuideOf: self, withInset: 20)
+        avatarView.autoHCenterInSuperview()
+        avatarView.autoSetDimension(.width, toSize: avatarSize)
+        avatarView.autoSetDimension(.height, toSize: avatarSize)
+
+        let nameLabel = UILabel()
+        nameLabel.text = contact.displayName
+        nameLabel.font = UIFont.ows_dynamicTypeTitle2.ows_bold()
+        nameLabel.textColor = UIColor.black
+        nameLabel.lineBreakMode = .byTruncatingTail
+        nameLabel.textAlignment = .center
+        topView.addSubview(nameLabel)
+        nameLabel.autoPinEdge(.top, to: .bottom, of: avatarView, withOffset: 10)
+        nameLabel.autoPinLeadingToSuperviewMargin(withInset: hMargin)
+        nameLabel.autoPinTrailingToSuperviewMargin(withInset: hMargin)
+
+        var lastView: UIView = nameLabel
```

Comment:
I find this `lastView` juggling, wherein it's a different value throughout the course of this layout pass to be a little confusing.

I guess we can't easily just use a stackview because we want custom spacing after each element...

Reply:
Yep, I've been trying to use `UIStackView` where I can.